ocaml/typing/typeopt.ml

```diff
@@ -298,21 +339,25 @@ let rec value_kind env ~loc ~visited ~depth ~num_nodes_visited ty
         let visited = Numbers.Int.Set.add (get_id ty) visited in
         match decl.type_kind with
         | Type_variant (cstrs, rep) ->
-          value_kind_variant env ~loc ~visited ~depth ~num_nodes_visited cstrs rep
+          fallback_if_missing_cmi (num_nodes_visited, Pgenval) (fun () ->
```

Comment:
I think this is wrong in the case of `[@@unboxed]` types. That is, if we have

```ocaml
type my_ufloat = float#   (* but this cmi is missing *)
type t = { x : my_ufloat } [@@unboxed]
let f (x : t) = x
```

we really don't want to return `Pgenval` when processing the arg of `f`.

Instead, I think it's better to have `fallback_if_missing_cmi` on the individual cases within `value_kind_record` and `value_kind_variant`.

Reply:
Of course, we don't actually support that type in this PR - a nice example of the tricky things I wanted to avoid dealing with in this PR :)

Your example is subtle.  I think to cause a problem all three lines need to be in separate files:
- If line 2 were in the file whose cmi is missing, we'd be unable to check `f`
- If line 2 were in the same file as `f`, we'd be unable to check the type declaration (because the cmi for `my_float` is missing.
So it must be the case that we compiled the first line, then compiled the second line with the cmi for the first, creating a `type_decl` for `t` (which has a layout in it), then compiled the third line without the cmi for the first.

What to do in this situation?  That is, even if we move the `fallback_if_missing_cmi` to the case of `value_kind_record` that deals with unboxed records, what should the fallback be, or should we just error?  I see four paths:

1. Just error (possibly with the nice cmi hint).  Reasonable.
2. Get the saved layout `l` off the record representation, and use `Typeopt.layout_of_sort (Layout.sort_of_layout l)` as the fallback.  Except we probably need a new version of `sort_of_layout` that fails more gracefully if the layout is `any`.  I propose this.
3. Compute the sort from the type of the field, and the (lambda) layout from that.  Also needs some graceful failure for `any`.  
4. Do nothing and leave a CR to deal with it when we support the type you are concerned about (like, next week, probably).

2 is unfortunate because it will fail "unnecessarily" in some corner cases where we couldn't compute a good layout while checking the declaration of `t`.  3 is unfortunate because we don't want to be doing this kind of layout computation in the translation to lambda (and have avoided it entirely - this would be the only case).

I guess I lean towards 2, but 4 is also tempting - especially since we won't be able to write a good test in this PR.

Reply:
Your comment about the lack of a test pushes me towards 4. But then next week, I advocate 5:

5. Take advantage of the value precondition. That is, we wouldn't be in any of this code unless the type that we're in is a value, so if we spot a missing cmi, we know what we're missing: `value` (or a sublayout thereof).

Actually, this suggests that the current code is just fine: even in the `[@@unboxed]` case, we know that `Pgenval` is an appropriate response, because the layout of the original type must be a `value` layout. So I withdraw my comment -- but maybe add a short comment in the code saying that `Pgenval` is always safe because of the precondition.